Mathlib/Algebra/Field/Defs.lean

```diff
@@ -47,61 +45,83 @@ a `GroupWithZero` lemma instead.
+
+If the division semiring has positive characteristic `p`, our division by zero convention forces
+`nnratCast (1 / p) = 1 / 0 = 0`. -/
+class DivisionSemiring (α : Type*) extends Semiring α, GroupWithZero α where
 #align division_semiring DivisionSemiring
 
 /-- A `DivisionRing` is a `Ring` with multiplicative inverses for nonzero elements.
 
 An instance of `DivisionRing K` includes maps `ratCast : ℚ → K` and `qsmul : ℚ → K → K`.
-If the division ring has positive characteristic p, we define `ratCast (1 / p) = 1 / 0 = 0`
-for consistency with our division by zero convention.
-The fields `ratCast` and `qsmul` are needed to implement the
-`algebraRat [DivisionRing K] : Algebra ℚ K` instance, since we need to control the specific
-definitions for some special cases of `K` (in particular `K = ℚ` itself).
-See also Note [forgetful inheritance].
--/
-class DivisionRing (K : Type u) extends Ring K, DivInvMonoid K, Nontrivial K, RatCast K where
+Those two fields are needed to implement the `DivisionRing K → Algebra ℚ K` instance since we need
+to control the specific definitions for some special cases of `K` (in particular `K = ℚ` itself).
+See also note [forgetful inheritance]. Similarly, there are maps `nnratCast ℚ≥0 → K` and
+`nnqsmul : ℚ≥0 → K → K` to implement the `DivisionSemiring K → Algebra ℚ≥0 K` instance.
+
+If the division ring has positive characteristic `p`, our division by zero convention forces
+`ratCast (1 / p) = 1 / 0 = 0`. -/
+class DivisionRing (α : Type*)
+  extends Ring α, DivInvMonoid α, Nontrivial α, RatCast α where
   /-- For a nonzero `a`, `a⁻¹` is a right multiplicative inverse. -/
-  protected mul_inv_cancel : ∀ (a : K), a ≠ 0 → a * a⁻¹ = 1
-  /-- We define the inverse of `0` to be `0`. -/
-  protected inv_zero : (0 : K)⁻¹ = 0
+  protected mul_inv_cancel : ∀ (a : α), a ≠ 0 → a * a⁻¹ = 1
+  /-- The inverse of `0` is `0` by convention. -/
+  protected inv_zero : (0 : α)⁻¹ = 0
   protected ratCast := Rat.castRec
-  /-- However `ratCast` is defined, propositionally it must be equal to `a * b⁻¹`. -/
-  protected ratCast_mk : ∀ (a : ℤ) (b : ℕ) (h1 h2), Rat.cast ⟨a, b, h1, h2⟩ = a * (b : K)⁻¹ := by
-    intros
-    rfl
-  /-- Multiplication by a rational number.
-  Set this to `qsmulRec _` unless there is a risk of a `Module ℚ _` instance diamond. -/
-  protected qsmul : ℚ → K → K
-  /-- However `qsmul` is defined,
-  propositionally it must be equal to multiplication by `ratCast`. -/
-  protected qsmul_eq_mul' : ∀ (a : ℚ) (x : K), qsmul a x = Rat.cast a * x := by
+  /-- However `Rat.cast` is defined, it must be equal to `a * b⁻¹`.
+
+  Do not use this lemma directly. Use `Rat.cast_def` instead. -/
+  protected ratCast_mk : ∀ (a : ℤ) (b : ℕ) (h1 h2), Rat.cast ⟨a, b, h1, h2⟩ = a * (b : α)⁻¹ := by
     intros
     rfl
+  /-- Scalar multiplication by a rational number.
+
+  Set this to `qsmulRec _` unless there is a risk of a `Module ℚ _` instance diamond.
+  Do not use directly. Instead use the `•` notation. -/
```

Comment:
I don't think we want to explicitly advocate for the default values.

Reply:
Nono, this is not advocating for default values. This is saying that people shouldn't be writing `DivisionRing.qsmul q a` but `q • a`

Reply:
No, above that it says, "Set this to `qsmulRec` unless ..."

Reply:
Oh, I didn't write that. @eric-wieser just added it in #11262